### DISCUSSION TITLE
Why are there so many #Unsupported for supported Expressions?

### DISCUSSION THREAD

**Discussion by manticore-projects** [prestodb/sql#35] [Q&A] [answered]
I would like to understand better, why so many productions return #Unsupported.
Example:
```sql
SELECT NEXT VALUE FOR a from b;
```

This valid statement is parsed perfectly fine, but instead of a `NEXT VALUE FOR` Node it would return an `Unsupported` Node. 
```
SQL Text
 └─DirectSqlStatement [1:1-1:31]
    └─Select [1:1-1:30]
       ├─SelectList [1:8-1:23]
       │  └─SelectItem [1:8-1:23]
       │     └─Unsupported [1:8-1:23]
       │        └─SchemaQualifiedName [1:23-1:23]
       │           └─Identifier: a [1:23-1:23]
       └─TableExpression [1:25-1:30]
          └─FromClause [1:25-1:30]
             └─TableName [1:30-1:30]
                └─Identifier: b [1:30-1:30]
```
Please care to elaborate the idea behind, I thank you a lot in advance.

**Comment by manticore-projects** [comment]
Being very curios, I just changed the production:
```
void next_value_expression() #NextValueFor:
{}
{
    "NEXT" "VALUE" "FOR" schema_qualified_name()
}
```

And as expected, I get an useful AST immediately:
```sql
SELECT NEXT VALUE FOR a from b;
```

```
SQL Text
 └─DirectSqlStatement [1:1-1:31]
    └─Select [1:1-1:30]
       ├─SelectList [1:8-1:23]
       │  └─SelectItem [1:8-1:23]
       │     └─NextValueFor [1:8-1:23]
       │        └─SchemaQualifiedName [1:23-1:23]
       │           └─Identifier: a [1:23-1:23]
       └─TableExpression [1:25-1:30]
          └─FromClause [1:25-1:30]
             └─TableName [1:30-1:30]
                └─Identifier: b [1:30-1:30]
```

**Comment by kaikalur** [comment, marked as answer]
We started with what Presto currently supports (for execution) only as real nodes and others as unsupported. In fact, I started with everything as unsupported and changed only those features supported by Presto as real nodes. That way when someone wants to implement a features, they can start with the grammar and go from there.

**Reply by manticore-projects** [reply]
Thank you, this makes it clear.
So would I be allowed to transform the `#Unsuppported` into real Nodes as long as a) they work for particular test cases and b) the work with the complete test suite?

**Reply by kaikalur** [reply]
Yeah, we can probably relax that now. See if you can make it conditional so that we give error for unsupported features if/when we use it in Presto.

Another option would be to add a validator class that has a hard-coded list of nodes that we currently have and any other node kind will be error out (and use that in the Presto version)
